File: Programmers/이분탐색/개똥벌레.py
'''
백준 3020번 - 개똥벌레
N : 동굴의 길이(짝수) (2 <= N <= 200,000)
H : 동굴의 높이 (2 <= H <= 500,000)
첫 번째 장애물은 항상 석순, 그 다음은 종유석과 석순이 번갈아가며 나타남
높이 1미터씩 구간을 나누고 구간을 지나갈 때 파괴해야하는 장애물의 최솟값과 그러한 구간이 총 몇개 있는지 출력
'''
from sys import stdin

n, h = map(int, input().split(' '))

path = [0 for _ in range(h)]    # 높이 만큼의 리스트

# 장애물(석순, 종유석)이 있는 위치를 표시
# 높이가 3인 석순이라면 path[0 ~ 2]의 값을 1씩 증가
# 높이가 4인 종유석이라면 ptrh[-1 ~ -4]의 값을 1씩 증가
for i in range(n) :
    obs = int(stdin.readline())
    if i % 2 == 0 :
        for j in range(obs) :
            path[j] += 1
    else :
        for j in range(h-1, h-obs-1, -1) :
            path[j] += 1

path.sort()
num = 0
min = path[0]
for i, v in enumerate(path) :
    if min != v :
        num = i
        break
print(min, num)
# min(path), path.count() 도 시간 초과

File: Programmers/이분탐색/개똥벌레.go
package 개똥벌레

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
)

func InputObs(n int) ([]int, []int) {
	/*
		https://jaewook.me/golang-fast-io-for-competitive-programming	Go Competitive Programming을 위한 빠른 입출력하기
	*/
	j := make([]int, 0, 10)
	s := make([]int, 0, 10)
	reader := bufio.NewReader(os.Stdin)

	for i := 0; i < n; i++ {
		bytes, _, _ := reader.ReadLine()
		num, _ := strconv.Atoi(string(bytes))

		if i%2 == 0 {
			j = append(j, num)
		} else {
			s = append(s, num)
		}
	}

	return j, s
}

func Bugs() {
	var n, h int
	fmt.Scan(&n, &h)

	j, s := InputObs(n)
	sort.Ints(j)
	sort.Ints(s)

	min := n
	count := 0

	// 익명함수 선언
	binarySearch := func(arr []int, target float32, start int, end int) int {
		var mid int
		for start <= end {
			mid = int((start + end) / 2)
			if float32(arr[mid]) < target {
				start = mid + 1
			} else {
				end = mid - 1
			}
		}
		return start
	}

	var j_count, s_count int
	for i := 1; i < h+1; i++ {
		j_count = len(j) - binarySearch(j, float32(i)-0.5, 0, len(j)-1)
		s_count = len(s) - binarySearch(s, float32(h-i)+0.5, 0, len(s)-1)

		if min == j_count+s_count {
			count += 1
		} else if min > j_count+s_count {
			count = 1
			min = j_count + s_count
		}
	}
	fmt.Println(min, count)
}

/*
제출 번호	  아이디	문제	 결과	       메모리	 시간	언어	코드 길이
35831696	lyw1217	3020	맞았습니다!!	7872	116	Go / 수정	1240
*/
